compiler/parser/src/lexer.rs

```diff
@@ -56,15 +71,89 @@ impl IndentationLevel {
     }
 }
 
+#[derive(Debug)]
+struct Indentations {
+    indent_stack: Vec<IndentationLevel>,
+}
+
+impl Indentations {
+    fn is_empty(&self) -> bool {
+        self.indent_stack.len() == 1
+    }
+
+    fn push(&mut self, indent: IndentationLevel) {
+        self.indent_stack.push(indent);
+    }
+
+    fn pop(&mut self) -> Option<IndentationLevel> {
+        if self.is_empty() {
+            return None;
+        }
+        self.indent_stack.pop()
+    }
+
+    fn current(&self) -> &IndentationLevel {
+        self.indent_stack
+            .last()
+            .expect("Indetations must have at least one level")
+    }
+}
+
+impl Default for Indentations {
+    fn default() -> Self {
+        Self {
+            indent_stack: vec![IndentationLevel::default()],
+        }
+    }
+}
+
+struct CharWindow<const N: usize>([Option<char>; N]);
+
+impl<const N: usize> CharWindow<N> {
+    fn slide(&mut self, next_char: Option<char>) {
+        let len = self.0.len();
+        for i in 0..(len - 1) {
+            self[i] = self[i + 1];
+        }
+        self[len - 1] = next_char;
+    }
+}
+
+impl<const N: usize> Default for CharWindow<N> {
+    fn default() -> Self {
+        Self([None; N])
+    }
+}
+
+impl<const N: usize, Idx> Index<Idx> for CharWindow<N>
+where
+    Idx: SliceIndex<[Option<char>], Output = Option<char>>,
+{
+    type Output = Option<char>;
+
+    fn index(&self, index: Idx) -> &Self::Output {
+        self.0.index(index)
+    }
+}
+
+impl<const N: usize, Idx> IndexMut<Idx> for CharWindow<N>
+where
+    Idx: SliceIndex<[Option<char>], Output = Option<char>>,
+{
+    fn index_mut(&mut self, index: Idx) -> &mut Self::Output {
+        self.0.index_mut(index)
+    }
+}
```

Comment:
```suggestion
impl<const N: usize> Default for CharWindow<N> {
    fn default() -> Self {
        Self([None; N])
    }
}
```
We will not need `IndexMut` by adding a proper method to CharWindow